kzg/src/msm/bos_coster.rs

```diff
@@ -0,0 +1,294 @@
+}
+
+impl<T> PartialEq for Pair<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.scalar == other.scalar
+    }
+}
+
+impl<T> Eq for Pair<T> {}
+
+impl<T> Ord for Pair<T> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.scalar.cmp(&other.scalar)
+    }
+}
+
+impl<T> PartialOrd for Pair<T> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<
+        TFr: Fr,
+        TG1Fp: G1Fp,
+        TG1: G1 + G1Mul<TFr> + G1GetFp<TG1Fp>,
+        TG1Affine: G1Affine<TG1, TG1Fp>,
+        TG1ProjAddAffine: G1ProjAddAffine<TG1, TG1Fp, TG1Affine>,
+    > BosCosterTable<TFr, TG1, TG1Fp, TG1Affine, TG1ProjAddAffine>
+{
+    pub fn new(points: &[TG1], matrix: &[Vec<TG1>]) -> Result<Option<Self>, String> {
+        if matrix.is_empty() {
+            Ok(Some(Self {
+                numpoints: points.len(),
+                points: Vec::from(points),
+
+                batch_numpoints: 0,
+                batch_points: Vec::new(),
+
+                fr_marker: PhantomData,
+                g1_fp_marker: PhantomData,
+                g1_affine_marker: PhantomData,
+                g1_affine_add_marker: PhantomData,
+            }))
+        } else {
+            Ok(Some(Self {
+                numpoints: points.len(),
+                points: Vec::from(points),
+
+                batch_numpoints: matrix[0].len(),
+                batch_points: Vec::from(matrix),
+
+                fr_marker: PhantomData,
+                g1_fp_marker: PhantomData,
+                g1_affine_marker: PhantomData,
+                g1_affine_add_marker: PhantomData,
+            }))
+        }
+    }
+
+    #[cfg(feature = "parallel")]
+    pub fn multiply_parallel(&self, scalars: &[TFr]) -> TG1 {
+        self.multiply_sequential(scalars)
+    }
+
+    pub fn multiply_sequential(&self, scalars: &[TFr]) -> TG1 {
+        Self::multiply_sequential_raw(&self.points, scalars)
+    }
+
+    // https://github.com/btclib-org/btclib was used as reference for this implementation
+    fn multiply_sequential_raw(bases: &[TG1], scalars: &[TFr]) -> TG1 {
+        let valid_pairs: Vec<Pair<TG1>> = scalars
+            .iter()
+            .zip(bases.iter())
+            .filter(|(s, p)| !s.is_zero() && !p.is_inf())
+            .map(|(s, p)| Pair {
+                scalar: TFr::to_scalar(&s),
+                point: p.clone(),
+            })
+            .collect();
+
+        let mut heap: BinaryHeap<Pair<TG1>> = BinaryHeap::from(valid_pairs);
+
+        while heap.len() > 1 {
+            let pair1: Pair<TG1> = heap.pop().unwrap();
+            let pair2_scalar = {
+                let mut pair2 = heap.peek_mut().unwrap();
+                let (k, pair2_scalar_2powk) = find_k(&pair1.scalar, &pair2.scalar);
+                let mut new_point = pair1.point.clone();
+                for _ in 0..k {
+                    new_point.dbl_assign();
+                }
+                pair2.point.add_assign(&new_point);
```

Comment:
unless you're absolutely sure that pair2 and new_point are not equal, you should use `add_or_dbl_assign`